### ISSUE TITLE
ClassCastException in cider middleware, when evaluating forms through an unix domain socket connection

### ISSUE THREAD

**Issue by bendlas** [clojure-emacs/cider-nrepl#727]
## Expected behavior

Connecting to nrepl and evaluating forms via unix domain socket connection works on latest JDK

## Actual behavior

After connecting successfully, evaluating forms throws `java.nio.HeapByteBuffer cannot be cast to class [Ljava.nio.ByteBuffer;`

```
Exception updating the ns-cache #error {
 :cause class java.nio.HeapByteBuffer cannot be cast to class [Ljava.nio.ByteBuffer; (java.nio.HeapByteBuffer and [Ljava.nio.ByteBuffer; are in module java.base of loader 'bootstrap')
 :via
 [{:type java.lang.ClassCastException
   :message class java.nio.HeapByteBuffer cannot be cast to class [Ljava.nio.ByteBuffer; (java.nio.HeapByteBuffer and [Ljava.nio.ByteBuffer; are in module java.base of loader 'bootstrap')
   :at [nrepl.socket.BufferedOutputChannel write socket.clj 207]}]
 :trace
 [[nrepl.socket.BufferedOutputChannel write socket.clj 207]
  [nrepl.socket.BufferedOutputChannel write socket.clj 202]
  [nrepl.transport$safe_write_bencode invokeStatic transport.clj 112]
  [nrepl.transport$safe_write_bencode invoke transport.clj 103]
  [nrepl.transport$bencode$fn__731 invoke transport.clj 130]
  [nrepl.transport.FnTransport send transport.clj 34]
  [nrepl.transport$eval637$fn__638$G__629__641 invoke transport.clj 23]
  [nrepl.transport$eval637$fn__638$G__628__645 invoke transport.clj 23]
  [clojure.lang.Var invoke Var.java 388]
  [cider.nrepl.middleware.track_state$update_and_send_cache$fn__7173 invoke track_state.clj 213]
  [cider.nrepl.middleware.track_state$update_and_send_cache invokeStatic track_state.clj 209]
  [cider.nrepl.middleware.track_state$update_and_send_cache invoke track_state.clj 159]
  [cider.nrepl.middleware.track_state$update_and_send_cache invokeStatic track_state.clj 181]
  [cider.nrepl.middleware.track_state$update_and_send_cache invoke track_state.clj 159]
  [clojure.lang.AFn applyToHelper AFn.java 156]
  [clojure.lang.AFn applyTo AFn.java 144]
  [clojure.core$apply invokeStatic core.clj 667]
  [clojure.core$update_in$up__6853 invoke core.clj 6185]
  [clojure.core$update_in invokeStatic core.clj 6186]
  [clojure.core$update_in doInvoke core.clj 6172]
  [clojure.lang.RestFn applyTo RestFn.java 146]
  [clojure.core$apply invokeStatic core.clj 671]
  [clojure.core$binding_conveyor_fn$fn__5754 doInvoke core.clj 2041]
  [clojure.lang.RestFn applyTo RestFn.java 146]
  [clojure.lang.Agent$Action doRun Agent.java 114]
  [clojure.lang.Agent$Action run Agent.java 163]
  [java.util.concurrent.ThreadPoolExecutor runWorker ThreadPoolExecutor.java 1136]
  [java.util.concurrent.ThreadPoolExecutor$Worker run ThreadPoolExecutor.java 635]
  [java.lang.Thread run Thread.java 833]]}
```

## Steps to reproduce the problem

- Create an nrepl server
  - on a unix domain file socket, with https://github.com/nrepl/nrepl/issues/245 fixed
  - with CIDER middleware
- Connect to created server
- Evaluate a form

I used the following method:

```sh
lein with-profile +nrepl run -m clojure.main -e "(do (require 'nrepl.server) (require 'cider.nrepl.middleware) (require 'refactor-nrepl.middleware) (nrepl.server/start-server :socket \"nrepl.sock\" :handler (apply nrepl.server/default-handler (conj cider.nrepl.middleware/cider-middleware #'refactor-nrepl.middleware/wrap-refactor))))"
```

Connect from emacs cider.el with https://github.com/clojure-emacs/cider/pull/3088

I couldn't figure out how to connect from `nrepl.cmdline` (sorry), but I'm reasonably confident that the problem is in the cider middlware, because when starting the server without middleware, the connection works.

## Environment & Version information

- `[nrepl/nrepl "0.9.0-beta4"]`
- `[cider/cider-nrepl "0.27.2"]`
- Clojure 1.10.1, Java 17.0.1

### cider-nrepl version

0.27.2

### Java version

17.0.1

### Operating system

NixOS Linux

**Comment by bendlas** [comment]
Note: on JDK11 with `com.kohlschutter.junixsocket/junixsocket-core`, this works as well.

**Comment by bbatsov** [comment]
Probably another wrong type annotation? @rldbv can you look into this?

**Comment by rlbdv** [comment]
Yep.

**Comment by rlbdv** [comment]
I'll see if I can set things up to reproduce the problem, but also wondered if we already knew what class was being passed as the `channel` in that socket.clj [`(.write ...)`](https://github.com/nrepl/nrepl/blob/ed6427d319b86c9045f4144c735e69b6fcc91ec1/src/clojure/nrepl/socket.clj#L207)?

**Comment by rlbdv** [comment]
@bendlas I was trying to follow your instructions, and the `lein run` command fails because `refactor-nrepl.middleware` is mssing.  I'm guessing either I don't have something set up right, or that's a local file?

**Comment by vemv** [comment]
I reckon you can change the command to

```
lein with-profile +nrepl run -m clojure.main -e "(do (require 'nrepl.server) (require 'cider.nrepl.middleware) (nrepl.server/start-server :socket \"nrepl.sock\" :handler (apply nrepl.server/default-handler cider.nrepl.middleware/cider-middleware)))"
```

i.e. with the refactor-nrepl middleware removed. refactor-nrepl almost certainly has nothing to do with the issue so it's safe to remove it.

**Comment by rlbdv** [comment]
Great - I can at least reproduce it.

**Comment by rlbdv** [comment]
Got it - should be able to post a fix tomorrow.

**Comment by bbatsov** [comment]
Great, thanks for the update!

**Comment by bbatsov** [comment]
This should be fixed in nREPL 0.9.0-beta5.
